### PR TITLE
Fix codegen to generate tests for nested classes into nested testclasses #560

### LINKED ISSUE #560: Inaccurate test naming for methods of nested classes
**Description**

Tests generated for methods of nested classes has the same names as if they were methods of enclosing class.

**To Reproduce**

Launch contest estimator with `projectFilter = null`, `methodFilter = "com.google.common.base.CaseFormat.*"` and `timeLimit = 120`. Look at tests generated for `CaseFormat.StringConverter.equals`

**Expected behavior**

Tests should be named something like `testStringConverterEqualsReturnsFalse` (or some other way to explicitly show that these are tests not for `CaseFormat.equals` but for `CaseFormat.StringConverter.equals`)

**Actual behavior**

Tests are named like `testEqualsReturnsFalse`

**Visual proofs (screenshots, logs, images)**

From generated testFile:

```Java
@Test
//@org.junit.jupiter.api.DisplayName("equals: object = Object() -> return false")
public void testEqualsReturnsFalse() throws Exception  {
    Object stringConverter = createInstance("com.google.common.base.CaseFormat$StringConverter");
    Object object = new Object();
    
    Class stringConverterClazz = Class.forName("com.google.common.base.CaseFormat$StringConverter");
    Class objectType = Class.forName("java.lang.Object");
    Method equalsMethod = stringConverterClazz.getDeclaredMethod("equals", objectType);
    equalsMethod.setAccessible(true);
    java.lang.Object[] equalsMethodArguments = new java.lang.Object[1];
    equalsMethodArguments[0] = object;
    boolean actual = ((Boolean) equalsMethod.invoke(stringConverter, equalsMethodArguments));
    
    assertFalse(actual);
}
```

**Environment**

Contest estimator, Java11

### PR DESCRIPTION
**Pull request by volivan239**
# Description

Changed codegen so that now structure of generated test classes corresponds to structure of classes in src code, i.e. tests generated for methods from inner/nested classes are now put to inner/nested classes in testfile.

Fixes #560

## Type of Change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

## Automated Testing

Added tests for ClassWithStaticAndInnerClasses, more tests to be added

## Manual Scenario 

Run on example described in #560 -- tests for `CaseFormat.StringConverter` are generated in nested class 

# Checklist (remove irrelevant options):

_This is the author self-check list_

- [ ] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [ ] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [ ] No new warnings
- [x] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

**Comment by volivan239** [comment]
In this PR some project tests were restructured. For example, `LombokAnnotationTest` had `testClass = EnumWithAnnotations::class` but, among others, contained tests for methods of `EnumWithoutAnnotations` and `NotNullAnnotations`, which caused incorrect behavior of codegen. Now this testclass is splitted.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/context/CgContext.kt [hunk 1]
```diff
@@ -262,7 +261,7 @@ internal interface CgContextOwner {
         }
     }
 
-    fun <R> withClassScope(block: () -> R): R {
+    fun <R> withTestClassFileScope(block: () -> R): R {
```
[the remark this answers is not in the document]
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/context/CgContext.kt]: should have been fixed in your commit

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/context/CgContext.kt [hunk 2]
```diff
@@ -307,6 +316,8 @@ internal interface CgContextOwner {
     }
 
     private fun clearClassScope() {
+        outerMostTestClassInfo.clear()
+        currentTestClassInfo = outerMostTestClassInfo
```
[the remark this answers is not in the document]
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/context/CgContext.kt]: same as above

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgTestClassConstructor.kt [hunk 3]
```diff
@@ -37,44 +41,87 @@ internal class CgTestClassConstructor(val context: CgContext) :
...
+            this.testClass = construct(testClass, true, testClassCustomName)
+            imports += context.collectedImports
+            testsGenerationReport = this@CgTestClassConstructor.testsGenerationReport
+        }
+    }
+
+    private fun construct(testClass: UtTestClass, outerMost: Boolean, customClassName: String? = null): CgTestClass {
+        return buildTestClass {
+            val (name, simpleName) = generateTestClassName(testClass, customClassName)
+            id = BuiltinClassId(
```
[the remark this answers is not in the document]
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgTestClassConstructor.kt]: added currentTestClass in addition to outerMostTestClass

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/TestClassInfo.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.utbot.framework.codegen.model.tree.CgAnnotation
+import org.utbot.framework.codegen.model.tree.CgMethod
+import org.utbot.framework.plugin.api.ClassId
+import org.utbot.framework.codegen.model.tree.CgTestClass
+
+/**
+ * This class stores context information needed to build [CgTestClass].
+ * Should only be used in [CgContextOwner].
+ */
+internal data class TestClassInfo(
```
[the remark this answers is not in the document]
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/TestClassInfo.kt]: done

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgTestClassConstructor.kt [hunk 5]
```diff
@@ -37,44 +41,87 @@ internal class CgTestClassConstructor(val context: CgContext) :
 
     private val methodConstructor = CgComponents.getMethodConstructorBy(context)
     private val nameGenerator = CgComponents.getNameGeneratorBy(context)
-
-    private val cgDataProviderMethods = mutableListOf<CgMethod>()
+    private val testFrameworkManager = CgComponents.getTestFrameworkManagerBy(context)
 
     private val testsGenerationReport: TestsGenerationReport = TestsGenerationReport()
 
+    private fun generateTestClassName(testClass: UtTestClass, customName: String? = null): Pair<String, String> {
```
[the remark this answers is not in the document]
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgTestClassConstructor.kt]: yes, after adding `currentTestClass` it is not needed, so removed it

File: utbot-framework/src/test/kotlin/org/utbot/examples/UtValueTestCaseChecker.kt [hunk 6]
```diff
@@ -2278,7 +2278,8 @@ abstract class UtValueTestCaseChecker(
     @Suppress("ControlFlowWithEmptyBody", "UNUSED_VARIABLE")
     private fun internalCheckForCodeGeneration(
         method: KFunction<*>,
-        testFrameworkConfiguration: TestFrameworkConfiguration
+        testFrameworkConfiguration: TestFrameworkConfiguration,
+        classUnderTest: KClass<*>? = null
```
[the remark this answers is not in the document]
Comment (ArsenHD) [reply utbot-framework/src/test/kotlin/org/utbot/examples/UtValueTestCaseChecker.kt]: Discussed and [reworked this method to be more readable](https://github.com/UnitTestBot/UTBotJava/pull/665/commits/be239e277bc835c98aca036594fadc48abf62d34).

**Review by ArsenHD — APPROVED** [review]

**Review by amandelpie — APPROVED** [review]

**Pull request merged**